kubernetes/updater/pkg/controller/updater.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/img"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/maintenance"
+	"github.com/gravitational/teleport/kubernetes/updater/pkg/version"
+)
+
+type VersionUpdater struct {
+	versionGetter       version.Getter
+	imageValidators     []img.Validator
+	maintenanceTriggers []maintenance.Trigger
+	baseImage           reference.Named
+}
+
+// TODO: find a better name
+func (r *VersionUpdater) Do(ctx context.Context, obj v1.Object, currentVersion string) (img.NamedTaggedDigested, error) {
+	// Those are debug logs only
+	log := ctrllog.FromContext(ctx).V(1)
+
+	// Can we do a maintenance?
+	log.Info("Checking if a maintenance trigger is on")
+	if !maintenance.AtLeastOne(ctx, obj, r.maintenanceTriggers...) {
+		return nil, &MaintenanceNotTriggeredError{}
+	}
+	log.Info("Maintenance triggered, getting new version")
+
+	// Get the next version
+	nextVersion, err := r.versionGetter.GetVersion()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Info("New version candidate", "nextVersion", nextVersion)
+	if !version.ValidVersionChange(currentVersion, nextVersion, "") {
+		return nil, &NoNewVersionError{CurrentVersion: currentVersion, NextVersion: nextVersion}
+	}
+
+	log.Info("Version change is valid, building img candidate")
+	// We tag our img candidate with the version
+	image, err := reference.WithTag(r.baseImage, nextVersion[1:])
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Info("Verifying candidate img", "img", image.String())
+	// We validate the signatures
+	valid, digested := img.AtLeastOneValid(ctx, image, r.imageValidators...)
+	if !valid {
+		return nil, trace.CompareFailed("no valid signature found for img %s", image)
+	}
+
+	log.Info("The following images were verified", "digestedImages", digested)
+	if len(digested) != 1 {
```

Comment:
What does this mean? That the same tag has two different digests?

Reply:
The tag has multiple valid signatures pointing to different digests. I don't think this it's possible right now with the [SimpleSigning](https://github.com/sigstore/cosign/blob/f9d8e5d22c038f261e6aa722af745d9bcc5850ea/specs/SIGNATURE_SPEC.md#simple-signing) but cosign validation functions are [returning a list of signatures](https://github.com/sigstore/cosign/blob/f9d8e5d22c038f261e6aa722af745d9bcc5850ea/pkg/cosign/verify.go#L471).

I was not able to build a working example so I'm not sure what to do there in such a weird state. I guess the sanest way would be to resolve the tag digest again, make sure at least 1 validated signature matches it, and use it.

It might be better to let the image validator deal with this complexity instead of the controller.

Reply:
I'm implementing the image signature validation and I confirm this is not possible to have multiple signatures with the same digest as long as we use `cosign.SimpleClaimValidator`. This seems like a reasonable hypothesis, worst case scenario we'll refactor this.

I'll change the function signature to only return a single `NamedTaggedDigested` image.